### PR TITLE
Fix packets deserialization on Mac

### PR DESCRIPTION
**Pull request by KasyanDiGris**
Fix of this bug: https://bugs.vcmi.eu/view.php?id=2970

### CODE REVIEW

File: client/CServerHandler.cpp [hunk 1]
```diff
@@ -39,6 +39,7 @@
 #include "../lib/registerTypes/RegisterTypes.h"
 #include "../lib/serializer/Connection.h"
 #include "../lib/serializer/CMemorySerializer.h"
+#include "../lib/serializer/StdInc.h"
```
Comment (alexvins) [inline client/CServerHandler.cpp]: "StdInc.h" name should be used only for global PCH

File: server/NetPacksServer.cpp [hunk 2]
```diff
@@ -7,7 +7,6 @@
  * Full text of license available in license.txt file, in main folder
  *
  */
-#include "StdInc.h"
```
Comment (alexvins) [inline server/NetPacksServer.cpp]: do not remove this line

File: lib/serializer/StdInc.cpp [hunk 3]
```diff
@@ -0,0 +1,2 @@
+// Creates the precompiled header
```
Comment (alexvins) [inline lib/serializer/StdInc.cpp]: Just delete this file

File: server/CVCMIServer.cpp [hunk 4]
```diff
@@ -7,8 +7,6 @@
  * Full text of license available in license.txt file, in main folder
  *
  */
-#include "StdInc.h"
```
Comment (alexvins) [inline server/CVCMIServer.cpp]: do not remove this line

File: server/CQuery.cpp [hunk 5]
```diff
@@ -7,11 +7,11 @@
  * Full text of license available in license.txt file, in main folder
  *
  */
-#include "StdInc.h"
```
Comment (alexvins) [inline server/CQuery.cpp]: do not remove this line

File: server/CGameHandler.cpp [hunk 6]
```diff
@@ -7,7 +7,7 @@
  * Full text of license available in license.txt file, in main folder
  *
  */
-#include "StdInc.h"
```
Comment (alexvins) [inline server/CGameHandler.cpp]: do not remove this line

File: lib/serializer/StdInc.h [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+#include <typeinfo>
+#include <string>
+#include "CTypeList.h"
+
+template<class T, class F>
+inline const T * dynamic_ptr_cast(const F * ptr)
+{
+	#ifndef __APPLE__
+  return dynamic_cast<const T*>(ptr);
```
Comment (alexvins) [inline lib/serializer/StdInc.h]: please format whole file according to guidelines, https://wiki.vcmi.eu/Coding_guidelines

File: lib/serializer/Cast.h [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+#pragma once
+
+#include <typeinfo>
+#include <string>
+#include "CTypeList.h"
+
+template<class T, class F>
+inline const T * dynamic_ptr_cast(const F * ptr)
+{
+#ifndef __APPLE__
```
Comment (alexvins) [inline lib/serializer/Cast.h]: use VCMI_APPLE define (same for non const overload)

**Comment by KasyanDiGris** [comment]
Can you please merge this PR

**Comment by ArseniyShestakov** [comment]
Ok I made some dumb cosmetic fixes and manually committed it here:
https://github.com/vcmi/vcmi/commit/96215233bc7248de3d6d55fac8f16d91fa6c84ad

Thanks for contribution!

**Pull request closed without merging**